diff --git a/src/Kernel/Support/XML.php b/src/Kernel/Support/XML.php
index b1318927..6e6fc1ae 100644
--- a/src/Kernel/Support/XML.php
+++ b/src/Kernel/Support/XML.php
@@ -27,11 +27,11 @@ class XML
      */
     public static function parse($xml)
     {
-        $backup = libxml_disable_entity_loader(true);
+        $backup = @libxml_disable_entity_loader(true);
 
         $result = self::normalize(simplexml_load_string(self::sanitize($xml), 'SimpleXMLElement', LIBXML_COMPACT | LIBXML_NOCDATA | LIBXML_NOBLANKS));
 
-        libxml_disable_entity_loader($backup);
+        @libxml_disable_entity_loader($backup);
 
         return $result;
     }
